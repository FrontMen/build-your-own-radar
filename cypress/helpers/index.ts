export const dataCy = (value: string) => cy.get(`[data-cy=${value}]`);
export const dataTestId = (value: string) => cy.get(`[data-testid=${value}]`);

export const hexToRgb = (hex: string) => {
  const bigint = parseInt(hex.replace('#', ''), 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;

  return `rgb(${r}, ${g}, ${b})`;
}
