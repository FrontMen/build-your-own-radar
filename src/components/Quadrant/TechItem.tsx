import React from 'react';
import styled from 'styled-components';
import { Typography } from 'src/Theme/Typography';
import { MediaQueries } from 'src/Theme/Helpers';
import { Link } from 'react-router-dom';
import { IoIosArrowRoundForward } from 'react-icons/io';

const ListItem = styled.li`
  list-style: none;
  width: 100%;
  break-inside: avoid;
  overflow: hidden;

  @media ${MediaQueries.phablet} {
    min-width: 9em;
  }
`;

const Details = styled.div<{ isOpened: boolean }>`
  max-height: ${props => (props.isOpened ? '75vh' : '0px')};
  transition: max-height 0.2s ease-in-out;
  overflow: hidden;
  background-color: ${props => props.theme.pallet.light};
`;
const Content = styled.div`
  padding: ${props => `${props.theme.space[2]}px`};
  overflow: hidden;
`;

const Label = styled.div<{ highlighted: boolean }>`
  ${Typography.body};
  cursor: pointer;
  padding: ${props => props.theme.space[1] / 2}px;
  color: ${props => props.highlighted && props.theme.pallet.white};
  background-color: ${props => props.highlighted && props.theme.pallet.blue};
  font-weight: ${props => props.highlighted && 600};
`;

const DetailsLink = styled(Link)`
  flex-direction: row;
  align-items: center;
  margin-top: ${props => props.theme.space[2]}px;
  width: 100%;
  color: ${props => props.theme.pallet.blue};
  display: flex;
  font-weight: 400;
  font-size: 0.9em;
  text-decoration: none;
`;

const ArrowRightIcon = styled(IoIosArrowRoundForward)`
  height: 2em;
  width: 2em;
`;

export interface TechnologyProps {
  technology: Technology;
  highlighted: string | null;
  setHighlighted: (a: string | null) => void;
  selected: string | null;
  setSelected: (a: string | null) => void;
  quadrant: string;
}

// use Component here to do smart memo which is much more tricky to achieve with hooks
export class TechItem extends React.Component<TechnologyProps> {
  shouldComponentUpdate(nextProps: TechnologyProps) {
    const {
      highlighted: prevHighlighted,
      selected: prevActive,
      technology: { name: prevName },
    } = this.props;

    const {
      highlighted: nextHighlighted,
      selected: nextActive,
      technology: { name: nextName },
    } = nextProps;

    // updating if highlighted property changed
    if (
      (prevHighlighted === prevName && nextHighlighted !== nextName) ||
      (prevHighlighted !== prevName && nextHighlighted === nextName)
    ) {
      return true;
    }
    // updating if selected property changed
    if (
      (prevActive === prevName && nextActive !== nextName) ||
      (prevActive !== prevName && nextActive === nextName)
    ) {
      return true;
    }
    // otherwise not
    return false;
  }

  handleMouseOver = () => this.props.setHighlighted(this.props.technology.name);

  handleMouseOut = () => this.props.setHighlighted(null);

  handleClick = () => {
    const {
      setSelected,
      technology: { name },
      selected,
    } = this.props;
    setSelected(name === selected ? null : name);
  };

  render() {
    const {
      selected,
      highlighted,
      technology: { name, description },
    } = this.props;

    const quadrantSlug = this.props.quadrant.toLowerCase();
    const technologySlug = name.toLowerCase();
    return (
      <ListItem>
        <Label
          data-testid="label"
          onMouseOver={this.handleMouseOver}
          onMouseOut={this.handleMouseOut}
          onClick={this.handleClick}
          highlighted={highlighted === name}
        >
          {name}
        </Label>

        <Details data-testid="details" isOpened={selected === name}>
          <Content>
            <span dangerouslySetInnerHTML={{ __html: description }} />
            <DetailsLink to={`/${quadrantSlug}/${technologySlug}`}>
              {'More '}
              <ArrowRightIcon data-testid="arrow-right-icon" />
            </DetailsLink>
          </Content>
        </Details>
      </ListItem>
    );
  }
}
