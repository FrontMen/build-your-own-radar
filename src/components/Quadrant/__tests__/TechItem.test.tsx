import React from 'react';
import { mount } from 'enzyme';

import { TechItem, TechnologyProps } from '../TechItem';
import { parsedMockDataItem } from 'test/mockData';
import { withAllProviders, AllProvidersWrapper } from 'test/helpers';
import { createMemoryHistory } from 'history';

const defaultProps: TechnologyProps = {
  highlighted: null,
  technology: parsedMockDataItem[0],
  selected: null,
  setSelected: jest.fn(),
  setHighlighted: jest.fn(),
  quadrant: 'tools',
};

const LabelSelector = {
  'data-testid': 'label',
};

const DetailsSelector = {
  'data-testid': 'details',
};

// used for tests which updates state, to not reinstantiate history obj
const history = createMemoryHistory({ initialEntries: ['/'] });

describe('TechItem', () => {
  it('it should matches the snapshot with default values', () => {
    const { container } = withAllProviders(<TechItem {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('it should highlight label when highlighted property matches current item', () => {
    const wrapper = mount(<TechItem {...defaultProps} />, {
      wrappingComponent: AllProvidersWrapper,
      wrappingComponentProps: { history },
    });

    let label = wrapper.find(LabelSelector);
    expect(label).toHaveStyleRule('font-weight', '400');

    wrapper.setProps({
      highlighted: parsedMockDataItem[0].positionId,
    });

    label = wrapper.find(LabelSelector);
    expect(label).toHaveStyleRule('font-weight', '600');
  });

  it('it should highlight label when selected property is set', () => {
    const wrapper = mount(<TechItem {...defaultProps} />, {
      wrappingComponent: AllProvidersWrapper,
      wrappingComponentProps: { history },
    });

    let label = wrapper.find(LabelSelector);
    expect(label).toHaveStyleRule('font-weight', '400');

    wrapper.setProps({
      selected: parsedMockDataItem[0].positionId,
    });

    label = wrapper.find(LabelSelector);
    expect(label).toHaveStyleRule('font-weight', '600');
  });

  it('it should expand details when selected prop matches item label', () => {
    const wrapper = mount<TechnologyProps>(<TechItem {...defaultProps} />, {
      wrappingComponent: AllProvidersWrapper,
      wrappingComponentProps: { history },
    });

    let details = wrapper.find(DetailsSelector);
    expect(details).toHaveStyleRule('max-height', '0px');

    wrapper.setProps({
      selected: parsedMockDataItem[0].positionId!,
    });

    details = wrapper.find(DetailsSelector);
    expect(details).toHaveStyleRule('max-height', '75vh');
  });

  it('should call setHighlighted callback on mouseOver and mouseOut on item Label', () => {
    const wrapper = mount(<TechItem {...defaultProps} />, {
      wrappingComponent: AllProvidersWrapper,
    });

    wrapper
      .find(LabelSelector)
      .at(0)
      .simulate('mouseOver');
    expect(defaultProps.setHighlighted).toHaveBeenCalledWith(
      parsedMockDataItem[0].positionId,
    );

    wrapper
      .find(LabelSelector)
      .at(0)
      .simulate('mouseOut');
    expect(defaultProps.setHighlighted).toHaveBeenCalledWith(null);
  });

  it('should call setActive callback on Label click', () => {
    const wrapper = mount(<TechItem {...defaultProps} />, {
      wrappingComponent: AllProvidersWrapper,
    });

    wrapper
      .find(LabelSelector)
      .at(0)
      .simulate('click');
    expect(defaultProps.setSelected).toHaveBeenCalledWith(
      `?tech=${parsedMockDataItem[0].positionId}`,
    );
  });
});
