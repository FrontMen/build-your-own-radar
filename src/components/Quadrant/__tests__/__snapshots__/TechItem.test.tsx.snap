// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TechItem it should matches the snapshot with default values 1`] = `
.c0 {
  list-style: none;
  width: 100%;
  -webkit-break-inside: avoid;
  break-inside: avoid;
  overflow: hidden;
}

.c2 {
  max-height: 0px;
  overflow: hidden;
  background-color: #f3f3f3;
}

.c3 {
  padding: 8px;
  overflow: hidden;
}

.c1 {
  font-family: Roboto,san-serif;
  font-size: 1em;
  font-weight: 400;
  line-height: 1.65;
  display: block;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  padding: 2px;
}

.c4 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 8px;
  width: 100%;
  color: #526de4;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: 400;
  font-size: 0.9em;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5 {
  height: 2em;
  width: 2em;
}

@media (min-width:481px) {
  .c0 {
    min-width: 9em;
  }
}

<div>
  <li
    class="c0"
    data-testid="list-item-Babel"
  >
    <div
      class="c1"
      data-testid="label"
    >
      Babel
    </div>
    <div
      class="c2"
      data-testid="details"
    >
      <div
        class="c3"
      >
        <span>
          This is the description. You can use basic html such as the 
          <strong>
            strong tag to emphasise keywords and phrases
          </strong>
           and insert
        </span>
        <a
          class="c4"
          href="/tools/babel"
        >
          More 
          <svg
            class="c5"
            data-testid="arrow-right-icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M295.6 163.7c-5.1 5-5.1 13.3-.1 18.4l60.8 60.9H124.9c-7.1 0-12.9 5.8-12.9 13s5.8 13 12.9 13h231.3l-60.8 60.9c-5 5.1-4.9 13.3.1 18.4 5.1 5 13.2 5 18.3-.1l82.4-83c1.1-1.2 2-2.5 2.7-4.1.7-1.6 1-3.3 1-5 0-3.4-1.3-6.6-3.7-9.1l-82.4-83c-4.9-5.2-13.1-5.3-18.2-.3z"
            />
          </svg>
        </a>
      </div>
    </div>
  </li>
</div>
`;
